#!/bin/env snakemake

picardfile = ['file']

infile = "{wcard}.dedup.bam"
infile_index = "{wcard}.dedup.bai"
outfile = "{wcard}.recal.bam"
####USER INPUTS####
capture_bed = "4"
###GATK FILES####
ref_genome = '/home/veera/faststorage/resources/reference_genome/human.b37.fasta'
gold_indel = '/data/refseq/tooldata/GATKbundle/2.8/b37/Mills_and_1000G_gold_standard.indels.b37.vcf'
dbsnp = '/data/refseq/tooldata/GATKbundle/2.8/b37/dbsnp_138.b37.vcf'

###GATK settings####
padding = "100"
######INTERMEDIATE FILES######
realign_targets = "{wcard}.realignment.targets.list"
realigned_bam = "{wcard}.realigned.bam"
realigned_bam_log = "{wcard}.realigned.bam.log"
recal_table = "{wcard}.recal.table"
post_recal_table = "{wcard}.post.recal.table"

rule gatk_pre_all:
    input: expand(outfile, wcard = picardfile)

rule gatk_realign_indel:
    input: a = infile, b = infile_index, r = ref_genome, gindel = gold_indel
    params: padding = padding, l = capture_bed
    output: temp(realign_targets)
    shell:
        "gatk -T RealignerTargetCreator "
        "-nt 4 "
        "-R {input.r} -L {params.l} --interval_padding {params.padding} "
        "-I {input.a} -known {input.gindel} "
        "-o {output}"

realign_command = "gatk -T IndelRealigner -R {input.r} -I {input.a} -targetIntervals {input.ti} -known {input.gindel} -o {output.a}"
rule gatk_realign_TI:
    input: a = infile, b = infile_index, ti = realign_targets, r = ref_genome, gindel = gold_indel
    output: a = temp(realigned_bam)
    params: ilog = temp(realigned_bam_log)
    run:
        shell(f"{realign_command} || {realign_command} -fixMisencodedQuals")
        

rule gatk_cov_pattern:
    input: a = realigned_bam, r = ref_genome, gindel = gold_indel, dbsnp = dbsnp
    output: temp(recal_table)
    shell:
        "gatk -T BaseRecalibrator "
        "-nct 4 -R {input.r} -I {input.a} -knownSites {input.dbsnp} -knownSites {input.gindel} "
        "-o {output} "

rule gatk_apply:
    input: a = infile, b = infile_index, r = ref_genome, recal = recal_table
    output: outfile
    shell:
        "gatk -T PrintReads "
        "-nct 4 -R {input.r} -I {input.a} -BQSR {input.recal} -o {output} "
